internal/core/src/indexbuilder/index_c.cpp

```diff
@@ -84,29 +84,90 @@ CreateIndexV0(enum CDataType dtype,
     return status;
 }
 
+milvus::storage::StorageConfig
+get_storage_config(const milvus::proto::indexcgo::StorageConfig& config) {
+    auto storage_config = milvus::storage::StorageConfig();
+    storage_config.address = std::string(config.address());
+    storage_config.bucket_name = std::string(config.bucket_name());
+    storage_config.access_key_id = std::string(config.access_keyid());
+    storage_config.access_key_value = std::string(config.secret_access_key());
+    storage_config.root_path = std::string(config.root_path());
+    storage_config.storage_type = std::string(config.storage_type());
+    storage_config.cloud_provider = std::string(config.cloud_provider());
+    storage_config.iam_endpoint = std::string(config.iamendpoint());
+    storage_config.cloud_provider = std::string(config.cloud_provider());
+    storage_config.useSSL = config.usessl();
+    storage_config.sslCACert = config.sslcacert();
+    storage_config.useIAM = config.useiam();
+    storage_config.region = config.region();
+    storage_config.useVirtualHost = config.use_virtual_host();
+    storage_config.requestTimeoutMs = config.request_timeout_ms();
+    return storage_config;
+}
+
+milvus::OptFieldT
+get_opt_field(const ::google::protobuf::RepeatedPtrField<
+              milvus::proto::indexcgo::OptionalFieldInfo>& field_infos) {
+    milvus::OptFieldT opt_fields_map;
+    for (const auto& field_info : field_infos) {
+        auto field_id = field_info.fieldid();
+        if (opt_fields_map.find(field_id) == opt_fields_map.end()) {
+            opt_fields_map[field_id] = {
+                field_info.field_name(),
+                static_cast<milvus::DataType>(field_info.field_type()),
+                {}};
+        }
+        for (const auto& str : field_info.data_paths()) {
+            std::get<2>(opt_fields_map[field_id]).emplace_back(str);
+        }
+    }
+
+    return opt_fields_map;
+}
+
+milvus::Config
+get_config(const milvus::proto::indexcgo::BuildIndexParams& info) {
+    milvus::Config config;
+    for (auto i = 0; i < info.index_params().size(); ++i) {
+        const auto& param = info.index_params(i);
+        config[param.key()] = param.value();
+    }
+
+    for (auto i = 0; i < info.type_params().size(); ++i) {
+        const auto& param = info.type_params(i);
+        config[param.key()] = param.value();
+    }
+
+    config["insert_files"] = info.insert_files();
+    if (info.opt_fields().size()) {
+        config["opt_fields"] = get_opt_field(info.opt_fields());
+    }
+
+    return config;
+}
+
 CStatus
-CreateIndex(CIndex* res_index, CBuildIndexInfo c_build_index_info) {
+CreateIndex(CIndex* res_index, const char* serialized_build_index_info) {
```

Comment:
1. We should use ParseFromArray/Marshal instead of ParseFromString/MarshalText.
2. The length of marshaled build index info should also be passed to avoid any blank character is contained.